File: internal/patterns/sourcesBackwards.go
package patterns

import (
	"io/fs"
	"os"
	"path"
	"slices"
	"strings"
)

// See [SourcesBackwards].
type SourcesBackwardsOptions struct {
	PatternFinderOptions
	Dir string
}

// Populates the [MatcherContext.External] map with [Source] objects.
func SourceBackwards(options SourcesBackwardsOptions) {
	dir := options.Dir
	for {
		if _, ok := options.Ctx.External[dir]; ok {
			break
		}

		foundSource := false

		for extractor := range slices.Values(options.Extractors) {
			var path string
			if dir == "." {
				path = extractor.Path
			} else {
				path = dir + "/" + extractor.Path
			}
			name := path[strings.LastIndexAny(path, "/")+1:]

			source := Source{
				Inverted: false,
				Name:     name,
				Path:     path,
				Pattern: SignedPattern{
					Exclude: Pattern{},
					Include: Pattern{},
				},
			}

			buff, err := fs.ReadFile(options.FS, path)
			if err != nil {
				if os.IsNotExist(err) {
					continue
				}
				source.Error = err
				options.Ctx.External[dir] = &source
				options.Ctx.Failed = true
				foundSource = true
			}

			options.Ctx.External[dir] = &source
			extractor.Extract(&source, buff, options.Ctx)
			if source.Error != nil {
				options.Ctx.Failed = true
				source.Error = err
				break
			}
			foundSource = true
			break
		}

		parent := path.Dir(dir)
		if !foundSource {
			if _, ok := options.Ctx.External[parent]; ok {
				options.Ctx.External[dir] = options.Ctx.External[parent]
			}
		}

		if dir == parent {
			break
		}
		dir = parent
	}

}

File: src/patterns/sourcesBackwards.ts
import { dirname } from "node:path"
// oxlint-disable-next-line no-unused-vars
import type { MatcherContext } from "./matcherContext.js"
import type { Source } from "./source.js"
import type { PatternFinderOptions } from "./extractor.js"
import { signedPatternCompile } from "./signedPattern.js"

/**
 * @see {@link sourcesBackwards}
 */
export interface SourcesBackwardsOptions extends PatternFinderOptions {
	dir: string
}

/**
 * Populates the {@link MatcherContext.external} map with {@link Source} objects.
 */
export async function sourcesBackwards(options: SourcesBackwardsOptions): Promise<void> {
	const { fs, ctx, cwd, extractors } = options
	let dir = options.dir

	while (true) {
		if (ctx.external.has(dir)) break

		let foundSource = false

		for (const extractor of extractors) {
			const path = dir === "." ? extractor.path : dir + "/" + extractor.path
			const name = path.substring(path.lastIndexOf("/") + 1)

			const source: Source = {
				inverted: false,
				name,
				path,
				pattern: { exclude: [], include: [], compiled: null },
			}

			let buff: Buffer | undefined
			try {
				buff = await fs.promises.readFile(cwd + "/" + path)
			} catch (err) {
				const error = err as NodeJS.ErrnoException
				if (error.code === "ENOENT") continue
				source.error = error
				ctx.external.set(dir, source)
				ctx.failed = true
				foundSource = true
				break
			}

			ctx.external.set(dir, source)
			try {
				extractor.extract(source, buff, ctx)
			} catch (err) {
				ctx.failed = true
				source.error =
					err instanceof Error
						? err
						: new Error("Unknown error during source extraction", { cause: err })
				break
			}
			signedPatternCompile(source.pattern)
			foundSource = true
			break
		}

		// Inherit from parent if not found
		const parent = dirname(dir)
		if (!foundSource) {
			if (ctx.external.has(parent)) {
				ctx.external.set(dir, ctx.external.get(parent)!)
			}
		}

		if (dir === parent) break
		dir = parent
	}
}
